Simplify code and make it more robust

H2CO3 opened the pull request:
# What does this PR do?

Make code more idiomatic, slightly more performant, less prone to some programming errors, and in other cases, simpler, using a couple of small changes:

- remove some unnecessary parentheses
- remove some unnecessary allocations (such as an identity vec-to-vec transform of the form `.into_iter().collect()`)
- replace manual impl of `next_multiple_of()` with the actual std function
- use `serde_json::to_vec()` instead of `to_string().into_bytes()`
- clean up some generic signatures
- extract the often-repeated magic number `8` into a named constant, `HEADER_SIZE_LEN`
- remove some redundant lifetime parameters and explicit annotations
- be explicit in comparing the size (and alignment) of data types when sorting tensors
- replace double negatives of the form `if negative_condition { Err(…) } else { Ok(…) }` with the condition un-negated and the branches flipped
- remove deep nesting and use early returns in some cases instead
- replace some instances of explicit size checking logic
- eta-reduce unnecessary closures: `|f| f.to_le_bytes()` -> `f32::to_le_bytes`
-

Narsil commented on safetensors/src/tensor.rs:
```diff
@@ -388,26 +400,26 @@ impl<'data> SafeTensors<'data> {
-        } else {
-            Err(SafeTensorError::TensorNotFound(tensor_name.to_string()))
-        }
+        let &Some(&index) = &self.metadata.index_map.get(tensor_name) else {
+            return Err(SafeTensorError::TensorNotFound(tensor_name.to_string()));
+        };
+
+        let Some(info) = &self.metadata.tensors.get(index) else {
+            return Err(SafeTensorError::TensorNotFound(tensor_name.to_string()));
+        };
+
+        Ok(TensorView {
+            dtype: info.dtype,
+            shape: info.shape.clone(),
+            data: &self.data[info.data_offsets.0..info.data_offsets.1],
+        })
```
This is purely style, right ? I personally find the first version more readable, the control flow here is really confusing with the let else syntax. Preferences...

Narsil commented on safetensors/src/tensor.rs:
```diff
@@ -388,26 +400,26 @@ impl<'data> SafeTensors<'data> {
-    pub fn names(&self) -> Vec<&'_ String> {
-        self.metadata.index_map.keys().collect()
+    pub fn names(&self) -> Vec<&'_ str> {
+        self.metadata.index_map.keys().map(String::as_str).collect()
```
I'm not sure I see the benefit here, callers can also call it themselves if they really want `str` instead of  `&String` no ?

H2CO3 replied:
The point here is that `&String` is *strictly* worse than `&str` because:
1. it exposes the implementation detail and forces the implementation to use _exactly_ a `String` (and not, e.g. a `Cow<str>` or an `Arc<str>`), while;
2. not having any benefits to the caller, because `&String` can't do anything more than `&str`. As it's immutable, any functionality that would be useful for `String` manipulation – such as appending characters or increasing the capacity – is inaccessible on the returned reference, so it provides only the functionality of `&str`.
As such, it's preferable to return just an `&str`, because it doesn't take away any functionality from the caller, while allowing the implementation to swap out the backing type if needed.

Narsil commented on safetensors/src/tensor.rs:
```diff
@@ -465,23 +477,21 @@ impl Serialize for Metadata {
-        for (name, info) in tensors {
-            map.serialize_entry(&name, &info)?;
+
+        for (name, info) in names.iter().zip(&self.tensors) {
```
This is much cleaner indeed.

Narsil commented on safetensors/src/tensor.rs:
```diff
@@ -507,7 +517,7 @@ impl Metadata {
             index_map,
         };
-        // metadata.validate()?;
+
```
While working on a secondary crate, I realized there was no public API for Metadata, so I marked this as pub in another PR.

Since it's now public we really need to keep validate (so I uncommented it), to make sure the Metadata can only be valid. (Distinc names, ordered data_offsets vectors etc..)

H2CO3 replied:
This doesn't remove the function itself, only the call. So are you saying the call itself is needed? (Otherwise, I would see no reason to keep around commented-out code.)

Narsil replied:
Yes, it was commented out as it was only an internal function, and at some point it was easier to remove the call on specific pathways to allow reading "malformed"  data (Metadata can follow the spec, but still not respect Metadata layout, which would cause issues down the line). While it wasn't public, we made sure the data was valid in other ways, since it's public we must reject any invalid data directly.

Narsil commented on safetensors/src/tensor.rs:
```diff
@@ -849,7 +863,7 @@ mod tests {
         let data: Vec<u8> = vec![0.0f32, 1.0, 2.0, 3.0, 4.0, 5.0]
             .into_iter()
-            .flat_map(|f| f.to_le_bytes())
+            .flat_map(f32::to_le_bytes)
```
I prefer the earlier version, this one is quite important, as you can keep the loop ISO event with different dtypes.

H2CO3 replied:
If you need to make this type-agnostic, the type can be left out and inference will fill it in, so `<_>::to_le_bytes` should work just as well.

Narsil commented on safetensors/src/tensor.rs:
```diff
@@ -989,9 +1003,10 @@ mod tests {
-        let mut tensors_desc = vec![];
-        tensors_desc.push(("wte".to_string(), vec![50257, 768]));
-        tensors_desc.push(("wpe".to_string(), vec![1024, 768]));
+        let mut tensors_desc = vec![
+            ("wte".to_string(), vec![50257, 768]),
+            ("wpe".to_string(), vec![1024, 768]),
+        ];
```
Thanks, this is cleaner indeed.

H2CO3 commented on safetensors/src/tensor.rs:
```diff
@@ -281,47 +294,49 @@ pub struct SafeTensors<'data> {
             .ok_or(SafeTensorError::InvalidHeaderLength)?;
-        if stop > buffer_len {
+
+        let Some(header_bytes) = buffer.get(HEADER_SIZE_LEN..stop) else {
```
> since purely slicing will return an incomplete slice instead.

No, that is incorrect. I think you are confusing this with Python's behavior. Rust does the correct and sensible thing and [returns `None` if either index is OOB](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=5ed6c35ae6f83e3dabe81f8cd13d4338).

Of course, feel free to add an explanatory comment anyway.

Narsil replied:
Indeed. I was thinking [..4] which does panic instead of returning None. for some reason I thought it was slicing inside.

H2CO3 commented on safetensors/src/tensor.rs:
```diff
@@ -281,47 +294,49 @@ pub struct SafeTensors<'data> {
             return Err(SafeTensorError::HeaderTooSmall);
-        }
-        let arr: [u8; 8] = [
-            buffer[0], buffer[1], buffer[2], buffer[3], buffer[4], buffer[5], buffer[6], buffer[7],
-        ];
+        };
+        let arr: [u8; HEADER_SIZE_LEN] = header_size_bytes
+            .try_into()
+            .map_err(|_| SafeTensorError::InvalidHeaderLength)?;
```
The array creation can't fail. The point is that this avoids the 8 individual bounds checks that the original code imposes.

If you prefer not returning an error, the `.map_err()` can be changed to `.unwrap()` or `.expect("this can't happen")`.

H2CO3 replied:
Yes, it can and probably will be eliminated indeed – this change is more about removing the repetitive code than raw performance.

H2CO3 commented on safetensors/src/tensor.rs:
```diff
@@ -197,16 +203,15 @@ fn prepare<S: AsRef<str> + Ord + core::fmt::Display, V: View, I: IntoIterator<It
-    // Force alignment to 8 bytes.
-    let extra = (8 - metadata_buf.len() % 8) % 8;
-    metadata_buf.extend(vec![b' '; extra]);
+    let mut metadata_buf = serde_json::to_vec(&metadata)?;
```
I don't understand what you are referring to. `serde_json::to_vec()` produces exactly the same output as `serde_json::to_string()`, it's just simpler than manually going through `to_string().into_bytes()` if you want a `Vec<u8>` anyway.

H2CO3 commented on safetensors/src/tensor.rs:
```diff
@@ -2,10 +2,12 @@
 use std::io::Write;
 
 const MAX_HEADER_SIZE: usize = 100_000_000;
+const HEADER_SIZE_LEN: usize = size_of::<u64>();
```
> This is not really HEADER_SIZE_LEN

Yes it is. I think you need to read and interpret the name more carefully. It _is_ the length of the header _size._

Note that I did **not** call it `HEADER_SIZE` or `HEADER_LEN`, because it's not the header size. It's the size of the number denoting the header length. Yes, it's 8 bytes, but hard-coding that constant a dozen times into the code is really just asking for trouble.

Narsil replied:
Let's go with N_LEN, it might be uninformative to you, but honestly keeping 8 everywhere hardcoded and called n when named is just fine to me. The whole point was to remove HEADER from the name.

> Yes, it's 8 bytes, but hard-coding that constant a dozen times into the code is really just asking for trouble.

Really ? I maintain quite a few librairies, and i don't think that it really matters. This constant will never change, and it's best descriptive name is "the first 8 bytes", so 8 isn't such a bad name.

> SIZEOF_HEADER_SIZE_NUMBER

I want the name to not contain HEADER, and let's keep it small, this is not java :)

H2CO3 commented on safetensors/src/tensor.rs:
```diff
@@ -168,22 +170,26 @@ pub trait View {
     let mut data: Vec<_> = data.into_iter().collect();
     data.sort_by(|(lname, left), (rname, right)| {
-        right.dtype().cmp(&left.dtype()).then(lname.cmp(rname))
+        right.dtype().size().cmp(&left.dtype().size()).then(lname.cmp(rname))
```
Sure, if that was the intention then feel free to discard this change. 👍  However, it would then be beneficial to somehow indicate this more clearly, since the current comment specifically says that it's about the alignment (which coincides with the size).

H2CO3 commented:
@Narsil I believe I've addressed all of your comments; I basically undone the changes where you explained why the original was necessary in all places except for the early returning of errors in `SafeTensors::tensor()`.
